Mathlib/Set.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura
+-/
+import Mathlib.Mem
+/-!
+
+# Sets
+
+This file sets up the theory of subsets of a type.
+
+## Main definitions
+
+* `{a : X | p a} : Set X` : the subset of `X` cut out by the predicate `p : X → Prop`
+* `{a | p a} : Set X` : equal to `{a : X | p a}`
+* `{a ∈ S | p a} : Set X` : the subset of `X` consisting of the elements of `S : set X`
+     which are cut out by `p`.
+
+## Implementation issues
+
+As in Lean 3, `Set X := X → Prop`
+
+I didn't call this file Data.Set.Basic because it contains core Lean 3
+stuff which happens before mathlib3's data.set.basic .
+This file is a port of the core Lean 3 file `lib/lean/library/init/data/set.lean`.
+
+## TODO
+
+Notation {a,b,c} for finite sets (both parser and prettyprinter).
+
+-/
+
+universes u v
+
+def Set (α : Type u) := α → Prop
+
+def setOf {α : Type u} (p : α → Prop) : Set α :=
+p
+
+class Subset (α : Type u) :=
+(subset : α → α → Prop)
+
+infix:50 " ⊆ " => Subset.subset
+
+class Union (α : Type u) :=
+(union : α → α → α)
+
+infixl:65 " ∪ " => Union.union
+
+class Inter (α : Type u) :=
+(inter : α → α → α)
+
+infixl:70 " ∩ " => Inter.inter
+
+class Sdiff (α : Type u) :=
+(sdiff : α → α → α)
+
+infix:70 " \\ " => Sdiff.sdiff
+
+namespace Set
+
+variable {α : Type u} {β : Type v}
+
+protected def mem (a : α) (s : Set α) :=
+s a
+
+instance : Mem α (Set α) :=
+⟨Set.mem⟩
+
+protected def subset (s₁ s₂ : Set α) :=
+∀ {a}, a ∈ s₁ → a ∈ s₂
+
+instance : Subset (Set α) :=
+⟨Set.subset⟩
+
+instance : EmptyCollection (Set α) :=
+⟨λ a => false⟩
+
+declare_syntax_cat binderterm -- notation for `a` or `a : A` or `a ∈ S`
+syntax ident : binderterm
+syntax ident ":" term : binderterm
+syntax ident "∈" term : binderterm
+
+-- Notation for sets
+syntax "{" binderterm "|" term "}" : term
```

Comment:
```suggestion
syntax ident " : " term : binderterm
syntax ident " ∈ " term : binderterm

-- Notation for sets
syntax "{ " binderterm " | " term " }" : term
```

It doesn't make a difference right now because the pretty-printer doesn't know about set comprehensions yet.  But the good news is that we can extend the pretty-printer ourselves, using so-called "delaborators".  Then these spaces make the difference between `{x∈s|x > 0}` and `{ x ∈ s | x > 0 }`.

Reply:
Thanks for the explanation.